src/uct/tcp/tcpcm/tcpcm_iface.c

```diff
@@ -0,0 +1,499 @@
+
+    iface_attr->iface_addr_len  = sizeof(ucs_sock_addr_t);
+    iface_attr->device_addr_len = 0;
+    iface_attr->cap.flags       = UCT_IFACE_FLAG_CONNECT_TO_SOCKADDR |
+                                  UCT_IFACE_FLAG_CB_ASYNC            |
+                                  UCT_IFACE_FLAG_ERRHANDLE_PEER_FAILURE;
+    /* User's private data size is UCT_TCPCM_UDP_PRIV_DATA_LEN minus room for
+     * the private_data header (to hold the length of the data) */
+    iface_attr->max_conn_priv   = UCT_TCPCM_MAX_CONN_PRIV;
+
+    return UCS_OK;
+}
+
+static int uct_tcpcm_iface_is_reachable(const uct_iface_h tl_iface,
+                                         const uct_device_addr_t *dev_addr,
+                                         const uct_iface_addr_t *iface_addr)
+{
+    /* Reachability can be checked with the uct_md_is_sockaddr_accessible API call */
+    return 1;
+}
+
+static ucs_status_t uct_tcpcm_iface_get_address(uct_iface_h tl_iface, uct_iface_addr_t *iface_addr)
+{
+    ucs_sock_addr_t *tcpcm_addr = (ucs_sock_addr_t *)iface_addr;
+
+    tcpcm_addr->addr    = NULL;
+    tcpcm_addr->addrlen = 0;
+    return UCS_OK;
+}
+
+#if 0
+static ucs_status_t uct_tcpcm_accept(int sock_id)
+{
+    /* The server will not send any reply data back to the client */
+    //struct tcp_conn_param conn_param = {0};
+
+    /* FIXME:
+    accept(int sockfd, struct sockaddr *addr, socklen_t *addrlen);
+    if (accept(id, &conn_param)) {
+        ucs_error("rdma_accept(to id=%p) failed: %m", id);
+        return UCS_ERR_IO_ERROR;
+    }
+    */
+
+    return UCS_OK;
+}
+#endif
+
+static ucs_status_t uct_tcpcm_iface_accept(uct_iface_h tl_iface,
+                                            uct_conn_request_h conn_request)
+{
+#if 0
+    struct sockaddr peer_addr;
+    socklen_t addrlen;
+    int accept_fd;
+    char ip_port_str[UCS_SOCKADDR_STRING_LEN];
+    ucs_status_t         status = UCS_OK;
+    uct_tcpcm_iface_t *iface = ucs_derived_of(tl_iface, uct_tcpcm_iface_t);
+
+    accept_fd = accept(iface->listen_fd, (struct sockaddr*)&peer_addr, &addrlen);
+    if (accept_fd < 0) {
+        if ((errno != EAGAIN) && (errno != EINTR)) {
+            ucs_error("accept() failed: %m");
+            return UCS_ERR_IO_ERROR;
+            // FIXME uct_tcp_iface_listen_close(iface);
+            //close(iface->sock_id);
+        }
+    }
+
+    ucs_debug("tcp_iface %p: accepted connection from %s at fd %d", iface,
+              ucs_sockaddr_str(&peer_addr, ip_port_str,
+                               UCS_SOCKADDR_STRING_LEN), accept_fd);
+
+    return status;
+#endif
+    return UCS_OK;
+}
+
+static ucs_status_t uct_tcpcm_iface_reject(uct_iface_h tl_iface,
+                                            uct_conn_request_h conn_request)
+{
+    // struct rdma_cm_event       *event = conn_request; FIXME
+    ucs_status_t               status = UCS_OK;
+    //uct_tcpcm_priv_data_hdr_t hdr    = {
+    //    .length = 0,
+    //    .status = UCS_ERR_REJECTED
+    //};
```

Comment:
```
shutdown();
close()
```
?

Reply:
Agree in principle. Given that `tcpcm_accept` is also a no-op at the moment (notice that `accept()` call is made at the event of connect request showing at the listening socket), I want to get clarification on this - Under what circumstances would `tcpcm_iface_reject` be called?

Reply:
UCP calls it when unable to handle Connection request w/o an error
or this can be called by UCP user through `ucp_listener_reject` API call - looks like it can be called to resolve simultaneous connections case (when a peer rejects a connection request from "loser" - and assumes that its connection should be accepted by "loser")